### PR TITLE
Feature/con6269 Refactor IDMatrix data mining function our of Application models - Gem

### PR DESCRIPTION
**Pull request by Tallmanjim**

### CODE REVIEW

File: lib/veda_idmatrix/request.rb [hunk 1]
```diff
@@ -84,15 +84,12 @@ def id_matrix_operation
 
     email_address = (self.entity[:email_address])
 
-    medicare_details = {}
-    unless self.entity[:medicare_card_number].blank?
```
Comment (elarex) [inline lib/veda_idmatrix/request.rb]: this used to guard against missing information, it no longer does.

File: spec/veda_idmatrix/request_spec.rb [hunk 2]
```diff
@@ -393,7 +393,9 @@
   describe 'medicare' do
     it 'should exclude if not defined' do
       no_mediacare = @entity_hash.dup
-      no_mediacare[:medicare_card_number] = ''
+      keys = [:medicare_card_number, :medicare_reference_number, :medicare_card_color, :medicare_card_expiry]
+      no_mediacare[keys.sample] = ''
+
       @request_no_medicare = VedaIdmatrix::Request.new(access: @access_hash, entity: no_mediacare, enquiry: @enquiry_hash)
       expect(@request_no_medicare.to_soap).to_not include('<idm:mediacare>')
```
Comment (elarex) [inline spec/veda_idmatrix/request_spec.rb]: mediacare ?
 MEDIA ? 
surely a typo that needs to be fixed.

**Comment by brentgreeff** [comment]
Its interesting - but I am not sure why I am watching this :)

File: lib/veda_idmatrix/request.rb [hunk 3]
```diff
@@ -113,14 +111,19 @@ def id_matrix_operation
...
+    #The search requires that the fields be present to be successful, otherwise we exclude the section
+    { 'drivers-licence-details' => drivers_licence_details, 'passport-details' => passport_details, 'medicare' => medicare_details}.each do |section, values|
+      details[:"#{section}"] = values unless self.mandatory_values_empty?(values)
+    end
 
     details
   end
 
+  def mandatory_values_empty?(values_hash)
+    values_hash.values.any? {|val| val.nil? || val.to_s.empty?}
```
Comment (elarex) [inline lib/veda_idmatrix/request.rb]: this is not valid syntax, you cannot pass a block to `any?`

File: lib/veda_idmatrix/request.rb [hunk 4]
```diff
@@ -84,15 +84,13 @@ def id_matrix_operation
...
     email_address = (self.entity[:email_address])
 
-    medicare_details = {}
+    medicare_details = {
+      :'card-number' => (self.entity[:medicare_card_number]),
+      :'reference-number' => (self.entity[:medicare_reference_number]),
+      :'card-colour' => (self.entity[:medicare_card_color]),
+      :'date-of-expiry' => (self.entity[:medicare_card_expiry])
+    }
     unless self.entity[:medicare_card_number].blank?
```
Comment (elarex) [inline lib/veda_idmatrix/request.rb]: this is not valid syntax, you cannot have an un-terminated conditional with no block.

File: spec/veda_idmatrix/request_spec.rb [hunk 5]
```diff
@@ -392,9 +392,11 @@
 
   describe 'medicare' do
     it 'should exclude if not defined' do
-      no_mediacare = @entity_hash.dup
-      no_mediacare[:medicare_card_number] = ''
-      @request_no_medicare = VedaIdmatrix::Request.new(access: @access_hash, entity: no_mediacare, enquiry: @enquiry_hash)
+      no_medicare = @entity_hash.dup
+      keys = [:medicare_card_number, :medicare_reference_number, :medicare_card_color, :medicare_card_expiry]
```
Comment (elarex) [inline spec/veda_idmatrix/request_spec.rb]: keys is defined, but apparently never used ...

File: lib/veda_idmatrix/request.rb [hunk 6]
```diff
@@ -113,14 +111,19 @@ def id_matrix_operation
     details[:'current-address'] = current_address
     details[:'phone']=  phone
     details[:'email-address'] = email_address
-    details[:'drivers-licence-details'] = drivers_licence_details
-    details[:'passport-details'] = passport_details
-    # Have to exclude medicare if details missing #5519
-    details[:'medicare'] = medicare_details unless medicare_details.empty?
+
```
Comment (elarex) [inline lib/veda_idmatrix/request.rb]: you cannot simply decide that all 3 of these sections should now be excluded when blank. They should always be present, with the exception of medicare, there is even a comment in the code about why this is the case.

**Pull request merged**